### ISSUE TITLE
[HOLD for payment 2023-06-02] [$1000] Switching between windows takes cell pointer to unpredictable position in the new magic code input field

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#18324]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Navigate to sign in page
2. Enter email & click on ‘Continue’ button
3. Open email in new window
4. Open magic code email
5. Open the window where expensify app is opened
6. Enter two or more magic code digits
7. Open the window where email is opened
8. Open the windows where expensify app is opened again
9. Enter more magic code digits

## Expected Result: 
Cell pointer moves to unpredictable position
## Actual Result: 
Cell pointer should remain at the position where we left of entering magic code

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.9.4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/235820473-69f835f6-7597-48eb-bdd4-bae853e467bc.mp4

https://user-images.githubusercontent.com/43996225/235820478-8094c7cb-c299-44f1-856b-0bb9d8df6d00.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @natnael-guchima
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1683073996655359

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~018aab7f6c72d1c7ad</li>
        <li>Upwork Job ID: 1657419055620501504</li>
        <li>Last Price Increase: 2023-05-13</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @alexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by alexpensify** [comment]
I'm catching up from being OOO on Friday and Monday, I'll try to test this one soon.

**Comment by alexpensify** [comment]
Thank you for flagging, I will review that GH.

**Comment by alexpensify** [comment]
Thank you @aimane-chnaif for the update. I'm putting this one on hold.

**Comment by alexpensify** [comment]
Still on hold

**Comment by alexpensify** [comment]
Still on hold

**Comment by Natnael-Guchima** [comment]
@alexpensify this https://github.com/Expensify/App/pull/18398 refractor seems to be completed and merged to staging. 
And the behavior still holds. 


https://github.com/Expensify/App/assets/57330233/86a487e6-d39c-45cf-8569-4ecd30075e02

cc: @aimane-chnaif

**Comment by alexpensify** [comment]
Thank you for the update @Natnael-Guchima - I'll test soon.

**Comment by alexpensify** [comment]
I tested on staging and was able to replicate the experience. The key here is to change tabs. Have staging open on one and then click another tab and then come back to the Expensify tab. I've taken this one off hold and assigned the `External` label.

**Comment by akinwale** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The magic code input does not behave as expected when the user switches away from the tab and switches back.

### What is the root cause of that problem?
When a user switches away from the validation page to another browser tab, the underlying `TextInput` component is blurred, which triggers the blur event. When the user switches back to the tab for the validation page, the validation page is focused. The `onFocus()` event is defined like so:

```
event.preventDefault();
setInput('');
```

This resets input value in the state to an empty string, and causes re-entry from the first item.

### What changes do you think we should make in order to solve the problem?
Remove the `setInput('');` call from the `onFocus` event, as this resets the state value.

### What alternative solutions did you explore? (Optional)
None.

**Comment by akinwale** [comment]
> @akinwale Do you know why doesn't this shift in the position happen always?

I'm not sure. In my testing on the current main branch, it always restarted from the beginning of the input. Maybe the refactor changed a few things? [This](https://github.com/Expensify/App/issues/18324#issuecomment-1544098365) is the behaviour that I was able to reliably reproduce.

**Comment by Santhosh-Sellavel** [comment]
@akinwale proposal looks good to me here!

cc: @francoisl

**Comment by alitoshmatov** [comment]
@Santhosh-Sellavel 
I think the root cause here is the fact that `onFocus` is being called on index 0, regardless of where were the last focus when switching tabs. One more observation is that when clicked on particular input, when tab switches the focus is re applied on that input rather that last active input. 

I think the objective here should be reapplying focus to the last active input when tab switches

You can see in console which input is getting onFocus event
https://github.com/Expensify/App/assets/59907218/1a4d93c1-dbe2-4ae5-a1cf-05f8ae0bf478

**Comment by Santhosh-Sellavel** [comment]
@alitoshmatov Interesting investigation. What do you propose here?

**Comment by alitoshmatov** [comment]
@Santhosh-Sellavel 
Interesting, and weird logic:
1. When we write we are still in first active input but highlight is changing looking like input also changed but it didn't
2. But when we delete, press, navigate with keyboard actual active input is changing, and when we write this input still being active

That's why when tab switches it is refocusing into last active input, because it looked like input changed when we wrote but in fact the active input didn't change but highlight it self changed

Notice that onChange event which tells current active input

Uploading magic.mov…

I propose:
1. Changing here `onChangeText` method to actually focus to next input, rather than just changing highlight itself. It would make much sense since everything else is changing active input
https://github.com/Expensify/App/blob/478de9c4dcc8d7ffa11de0f29df9cda53f2ab7ba/src/components/MagicCodeInput.js#L190-L191
to this
```javascript
inputRefs.current[updatedFocusedIndex].focus();
        // setFocusedIndex(updatedFocusedIndex); // no need because we are going to handle it in onFocus
        // setInput(value); //same
``` 

2. Change `onFocus` handler to also change highlight
https://github.com/Expensify/App/blob/478de9c4dcc8d7ffa11de0f29df9cda53f2ab7ba/src/components/MagicCodeInput.js#L146-L149
to this
```javascript
    const onFocus = (event, index) => {
        event.preventDefault();
        setInput('');
        setFocusedIndex(index);
        setEditIndex(index);
    };

``` 

I tested the code and it is producing the same behavior and fixing switching tab issue

**Comment by akinwale** [comment]
@francoisl Just checking in here to see if we're moving forward with this. Thanks.

**Comment by Santhosh-Sellavel** [comment]
@francoisl Please hold on, as there might be a better solution with a root cause [here](https://github.com/Expensify/App/issues/18324#issuecomment-1548367474) will review and update soon

**Comment by francoisl** [comment]
Yes sounds good, I was just reviewing and testing the two solutions. Both work for me so far.

One small note about the second solution is that the `onPress` and `onFocus` functions would now be identical, we can try to consolidate them into one function to avoid repetition.

Still curious if you have a preference for one of the solutions @Santhosh-Sellavel.

**Comment by alexpensify** [comment]
@dylanexpensify - assigning if a BZ team member is needed to step in while I'm OOO.

**Comment by Santhosh-Sellavel** [comment]
@alitoshmatov It's not working for me after switching tabs.

https://github.com/Expensify/App/assets/85645967/fc1f2c3a-44dd-4ea7-8398-c4fd12bcba8c

**Comment by Santhosh-Sellavel** [comment]
Also @akinwale proposal will create this deploy blockler regression

https://github.com/Expensify/App/assets/85645967/700f37a5-836c-4164-86ba-3952a29c80e8

cc: @francoisl

**Comment by francoisl** [comment]
@Santhosh-Sellavel just to make extra sure, did you also update the `onFocus={onFocus}` line? Here's the full diff I used and it works for me:

```diff
diff --git a/src/components/MagicCodeInput.js b/src/components/MagicCodeInput.js
index 2d4a31fde8..a343d86e8d 100644
--- a/src/components/MagicCodeInput.js
+++ b/src/components/MagicCodeInput.js
@@ -144,10 +144,12 @@ function MagicCodeInput(props) {
      * @param {Object} event
      * @param {Number} index
      */
-    const onFocus = (event) => {
+    const onFocus = (event, index) => {
         event.preventDefault();
         setInput('');
-    };
+        setFocusedIndex(index);
+        setEditIndex(index);
+    }
 
     /**
      * Callback for the onPress event, updates the indexes
@@ -188,8 +190,9 @@ function MagicCodeInput(props) {
         let numbers = decomposeString(props.value);
         numbers = [...numbers.slice(0, editIndex), ...numbersArr, ...numbers.slice(numbersArr.length + editIndex, CONST.MAGIC_CODE_LENGTH)];
 
-        setFocusedIndex(updatedFocusedIndex);
-        setInput(value);
+        inputRefs.current[updatedFocusedIndex].focus();
+        // setFocusedIndex(updatedFocusedIndex);
+        // setInput(value);
 
         const finalInput = composeToString(numbers);
         props.onChangeText(finalInput);
@@ -300,7 +303,7 @@ function MagicCodeInput(props) {
                                 }}
                                 onKeyPress={onKeyPress}
                                 onPress={(event) => onPress(event, index)}
-                                onFocus={onFocus}
+                                onFocus={(event) => onFocus(event, index)}
                                 caretHidden={isMobileSafari}
                                 inputStyle={[isMobileSafari ? styles.magicCodeInputTransparent : undefined]}
                             />
```

Looking into that regression you noticed now.

**Comment by Santhosh-Sellavel** [comment]
🤦 Ah thanks @francoisl, I was wondering how it was working yesterday, missed it today in the rush!

Yeah, it's working let's go with @alitoshmatov's proposal thanks!

**Comment by akinwale** [comment]
> Also, @akinwale proposal will create this deploy blocker regression
> 
>  Screen.Recording.2023-05-18.at.12.49.16.AM.mov 
> cc: @francoisl

This is not possible with `setInput('');` completely removed, which is the approach that I ultimately recommended in my proposal as highlighted by the note. The `onFocus` event handler should look like this:
```
const onFocus = (event) => {
    event.preventDefault();
};
```

**Comment by Santhosh-Sellavel** [comment]
@akinwale You updated the proposal with a note later which I'm unaware of until now. Your initial proposal didn't have a note or didn't mention the issue. Check out this [comment](https://github.com/Expensify/App/issues/18324#issuecomment-1546717889).

Here you suggested replacing the place `setInput('');` with `setInput(!_.isEmpty(props.value) ? props.value.trim() : '');` and unsure here why about it's happening. Initially, I didn't test the other flow, just tried to reproduce the original issue which was fixed but it broke the delete password case.

cc: @francoisl

**Comment by akinwale** [comment]
> @akinwale You updated the proposal with a note later which I'm unaware of until now. Your initial proposal didn't have a note or didn't mention the issue. Check out this [comment](https://github.com/Expensify/App/issues/18324#issuecomment-1546717889).

That's a quoted comment of my original proposal, and I would have expected that a review would be done on my final proposal. I also got confused at some point by the quoted proposal, so that's understandable, but all the edits on my proposal were completed before you conducted your initial review. Either way, the first thing that the proposal states is to remove `setInput('')`. I simply added the note to provide some context and clarification.

> Here you suggested replacing the place `setInput('');` with `setInput(!_.isEmpty(props.value) ? props.value.trim() : '');` and unsure here why about it's happening. Initially, I didn't test the other flow, just tried to reproduce the original issue which was fixed but it broke the delete password case.
> 
> cc: @francoisl

I suggested conditionally replacing the `setInput` and I provided additional clarification that this will cause some issues that will require some refactoring, hence my included note.

> If the setInput method absolutely has to be called in the onFocus event handler for whatever reason, then we can use it like so (by taking into consideration the value of the user's previous input):

I don't know what the delete password case means here. Are you referring to pressing Backspace on the magic code input? If this is the case, then yes, this is one of the caveats that I raised in my proposal, and why I included a note that we should stick with just removing the `setInput('')` call.

**Comment by Santhosh-Sellavel** [comment]
Sorry, It's a delete magic code case this [one](https://github.com/Expensify/App/issues/18324#issuecomment-1551930783).

> but all the edits on my proposal were completed before you conducted your initial review.

Yeah, it seems like. I don't remember seeing it either. Maybe this GitHub platform issue I must have opened it might not have been updated for edits as it does for new comments.

**Comment by francoisl** [comment]
One thing we didn't notice with the second proposal is that if we no longer need the line `setInput(value)`, then all the remaining uses of `setInput` are `setInput('')`, meaning that the `input` state never changes, and we don't even need it anymore?

**Comment by Santhosh-Sellavel** [comment]
@francoisl 
> One thing we didn't notice with the second proposal is that if we no longer need the line setInput(value), then all the remaining uses of setInput are setInput(''), meaning that the input state never changes, and we don't even need it anymore?

Yeah, it's kind of weird, but we should be sure before removing it. Thoughts @alitoshmatov ?

 Maybe the component is up for a refactor.

**Comment by francoisl** [comment]
The component is pretty recent, so hopefully it doesn't need a refactor already. But yes, let's see if there's a further possible simplification with @alitoshmatov's proposal, otherwise let's go with @akinwale's proposal which still needs the `input` state.

**Comment by alitoshmatov** [comment]
@francoisl @Santhosh-Sellavel 
Yes it looks like we don't need to store input value. Because with this approach, we are making sure that every input is responsible for just itself, handles focus, onPress of itself and produces only single digit. And we are keeping track of of which input is being edited with `editIndex`.

**Comment by alitoshmatov** [comment]
One correction about my last statement

> we are making sure that every input is responsible for just itself, handles focus, onPress of itself and **produces only single digit**

Above statement is not true for pasting value with multiple digits, and it is handled by existing logic in `onChangeText` function

**Comment by francoisl** [comment]
@alitoshmatov for the sake of clarity, can you post an updated proposal please? I'm not super clear on your comment about `onChangeText` and just want to make sure we're on the same page. Thanks!

**Comment by alitoshmatov** [comment]
@francoisl

## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Switching between windows takes the cell pointer to an unpredictable position in the new magic code input field 
### What is the root cause of that problem?
When we write we are still in the first active input but the highlight is changing looking like the input also changed but it didn't
But when we delete, press, and navigate with the keyboard actual active input is changing, and when we write this input still being active
That's why when the tab switches it refocuses into last active input, because it looked like the input changed when we wrote but in fact, the active input didn't change but highlighted itself changed
### What changes do you think we should make in order to solve the problem?
We should ensure that the view user sees and underlying inputs should match, meaning that the corresponding input should also be active whatever number slot is highlighted.

To achieve this we should focus on the next input when change happens. Change following
https://github.com/Expensify/App/blob/478de9c4dcc8d7ffa11de0f29df9cda53f2ab7ba/src/components/MagicCodeInput.js#L190-L191
to this
```javascript
inputRefs.current[updatedFocusedIndex].focus();
        // setFocusedIndex(updatedFocusedIndex); // no need because we are going to handle it in onFocus
        // setInput(value); //same
``` 
To keep the view and the input in sync, we should update the highlighted view when onFocus happens, meaning that whichever input is active the corresponding view input is also highlighted
Change the `onFocus` handler to also change the highlight
https://github.com/Expensify/App/blob/478de9c4dcc8d7ffa11de0f29df9cda53f2ab7ba/src/components/MagicCodeInput.js#L146-L149

```javascript
    const onFocus = (event, index) => {
        event.preventDefault();
        setFocusedIndex(index);
        setEditIndex(index);
    };

``` 

Since we no longer need input value stored in the state we can remove all instances of `setInput` and remove this
https://github.com/Expensify/App/blob/478de9c4dcc8d7ffa11de0f29df9cda53f2ab7ba/src/components/MagicCodeInput.js#L281

We only rely on the state provided by the parent which is `props.value` and change it with `props.onChangeText`

### What alternative solutions did you explore? (Optional)

**Comment by alitoshmatov** [comment]
@francoisl 
I think it would be great if we could get an opinion of the developer who worked on building this

**Comment by francoisl** [comment]
Good idea, started a discussion in Slack [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1684530493448609).

**Comment by dylanexpensify** [comment]
Bumped the convo!

**Comment by dylanexpensify** [comment]
[I'm heading OOO tomorrow](https://expensify.slack.com/archives/C01SKUP7QR0/p1684776773511559) for about a week so adding another BZ member while I'm away to help keep the train moving! 🚂  Thanks @greg-schroeder!!!

**Comment by greg-schroeder** [comment]
I gotchu

**Comment by alexpensify** [comment]
@greg-schroeder and @dylanexpensify - I appreciate your help here! I'm back online and removing your assignment.

**Comment by alexpensify** [comment]
@Santhosh-Sellavel - When you get a chance, can you please fill out the checklist: https://github.com/Expensify/App/issues/18324#issuecomment-1563877934

Thanks!

**Comment by alexpensify** [comment]
@Santhosh-Sellavel, @akinwale, @Natnael-Guchima 

To prepare for the payment date, please apply to the job in Upwork: 

https://www.upwork.com/jobs/~018aab7f6c72d1c7ad

Thanks!

**Comment by akinwale** [comment]
> @Santhosh-Sellavel, @akinwale, @Natnael-Guchima
> 
> To prepare for the payment date, please apply to the job in Upwork:
> 
> https://www.upwork.com/jobs/~018aab7f6c72d1c7ad
> 
> Thanks!

Already applied. Thanks.

**Comment by Natnael-Guchima** [comment]
> @Santhosh-Sellavel, @akinwale, @Natnael-Guchima 
> 
> To prepare for the payment date, please apply to the job in Upwork: 
> 
> https://www.upwork.com/jobs/~018aab7f6c72d1c7ad
> 
> Thanks! 


Applied. Thanks!

**Comment by alexpensify** [comment]
@Santhosh-Sellavel - bumping this one, can you please apply to the Upwork job? I want to prepare the payments for June 2. Thanks!

**Comment by alexpensify** [comment]
Everyone else is party-ready in Upwork, so I see you @Natnael-Guchima and @akinwale.

**Comment by alexpensify** [comment]
@Santhosh-Sellavel, @akinwale, @Natnael-Guchima - the payments have been prepared in Upwork. Please accept and I can complete the process.

**Comment by akinwale** [comment]
> @Santhosh-Sellavel, @akinwale, @Natnael-Guchima - the payments have been prepared in Upwork. Please accept and I can complete the process.

@alexpensify Accepted, thanks!

**Comment by Santhosh-Sellavel** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
> 
> * [x]  [@Santhosh-Sellavel] The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/16076/files#r1215047427
> * [x]  [@Santhosh-Sellavel] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/16076/files#r1215047427
> * [x]  [@Santhosh-Sellavel] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: N/A I think should be caught during the PR testing
> * [x]  [@Santhosh-Sellavel] Determine if we should create a regression test for this bug. N/A That's the normal expection
> * [x]  [@Santhosh-Sellavel] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
> * [ ]  [@alexpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by Santhosh-Sellavel** [comment]
### Regression Test Steps 

1. Navigate to the sign-in page.
2. Enter your email and click the ‘Continue’ button.
3. Open a new browser tab and navigate to any web page.
4. Switch to the browser tab with the magic code and enter the first 2 or 3 digits of the magic code.
5. Switch to the other browser tab.
6. Switch back to the browser tab with the magic code input and attempt to enter more digits.
7. Verify that the new digits entered continue from the current point, instead of overwriting the first set of digits that were already entered.

👍 or 👎

**Comment by Natnael-Guchima** [comment]
> @Santhosh-Sellavel, @akinwale, @Natnael-Guchima - the payments have been prepared in Upwork. Please accept and I can complete the process. 

Accepted! Thanks, @alexpensify.

**Comment by alexpensify** [comment]
@Santhosh-Sellavel, @akinwale, @Natnael-Guchima - everyone has been paid in Upwork. I'll wait to confirm again on Monday, then will close the GH.

**Comment by akinwale** [comment]
> @Santhosh-Sellavel, @akinwale, @Natnael-Guchima - everyone has been paid in Upwork. I'll wait to confirm again on Monday, then will close the GH.

Payment received. Thanks!

**Comment by Santhosh-Sellavel** [comment]
Got it thanks @alexpensify!

**Comment by Natnael-Guchima** [comment]
> @Santhosh-Sellavel, @akinwale, @Natnael-Guchima - everyone has been paid in Upwork. I'll wait to confirm again on Monday, then will close the GH.

Payment received.  Thanks!

**Comment by alexpensify** [comment]
Thank you for confirming! I'm closing this GH and the job in Upwork. Great work here!
